src/Microsoft.Net.Http.Headers/HeaderUtilities.cs

```diff
@@ -594,13 +595,64 @@ public static string FormatDate(DateTimeOffset dateTime, bool quoted)
 
         public static StringSegment RemoveQuotes(StringSegment input)
         {
-            if (!StringSegment.IsNullOrEmpty(input) && input.Length >= 2 && input[0] == '"' && input[input.Length - 1] == '"')
+            if (IsQuoted(input))
             {
                 input = input.Subsegment(1, input.Length - 2);
             }
             return input;
         }
 
+        public static bool IsQuoted(StringSegment input)
+        {
+            return !StringSegment.IsNullOrEmpty(input) && input.Length >= 2 && input[0] == '"' && input[input.Length - 1] == '"';
+        }
+
+        // See https://tools.ietf.org/html/rfc2616#section-2.2
+        public static StringSegment DecodeQuotedString(StringSegment input)
+        {
+            input = RemoveQuotes(input);
+            var stringBuilder = new StringBuilder(input.Length);
+
+            if (input.IndexOf('\\') == -1)
+            {
+                return input;
+            }
+
+            for (var i = 0; i < input.Length; i++)
+            {
+                if (input[i] == '\\')
+                {
+                    if (input[i + 1] == '\\' || input[i + 1] == '\"')
+                    {
+                        stringBuilder.Append(input[i + 1]);
+                        i++;
+                        continue;
+                    }
+                }
+                stringBuilder.Append(input[i]);
+            }
+
+            return stringBuilder.ToString();
+        }
+
+        // See https://tools.ietf.org/html/rfc2616#section-2.2
+        public static StringSegment EscapeAsQuotedString(StringSegment input)
+        {
+            var stringBuilder = new StringBuilder(input.Length * 2);  // worst case the string size doubles as every character must be escaped.
```

Comment:
- Do not allocate if `input` contains only token characters i.e. if the quotes aren't needed.
- Should the StringBuilder instances be pooled somewhere? They can be large and are reusable.
- If not pooling, make the `StringBuilder` as small as possible by calculating how many backslashes must be added before creating the instance.
- `input.Length*2` is two characters too short for the worst case. Don't forget about the surrounding quotes.